File: LeetCode/112_PathSum.py
# https://leetcode.com/problems/path-sum/
# 2021 12.12


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        # # v1 copied Iterative using stack
        # if not root:
        #     return False
        # stack=[(root,root.val)]
        # while stack:
        #     curr,val=stack.pop()
        #     if not curr.left and not curr.right and val==targetSum:
        #         return True
        #     if curr.left:
        #         stack.append((curr.left,val+curr.left.val))
        #     if curr.right:
        #         stack.append((curr.right,val+curr.right.val))
        # return False

        # v2 copied Recursive, neat and birlliant
        if not root:
            return False
        if not root.left and not root.right and root.val==targetSum:
            return True
        targetSum-=root.val
        return self.hasPathSum(root.left,targetSum) or self.hasPathSum(root.right,targetSum)

File: LeetCode/112_PathSum.java
// https://leetcode.com/problems/path-sum/submissions/
// 2022年09月24日 11:31:57


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean hasPathSum(TreeNode root, int targetSum) {
        if(root==null){
            return false;
        }
        if (root.left==null && root.right==null && root.val==targetSum){
            return true;
        }
        targetSum-=root.val;
        return hasPathSum(root.left,targetSum) || hasPathSum(root.right,targetSum);
        
    }
}

// ###### 2023年08月24日 14:50:31
class Solution {
    public boolean hasPathSum(TreeNode root, int targetSum) {
        if(root==null){return false;}
        if(root.left==null && root.right==null & root.val==targetSum){return true;}
        return hasPathSum(root.left,targetSum-root.val)||hasPathSum(root.right,targetSum-root.val);
        }
}
    
